### PR TITLE
DOCKER-788: update image/container filtering for new API format

### PR DESCRIPTION
**Pull request by twhiteman**
With Docker 1.10 (API version 1.22) the filters changed from an array to an
object with the value true.

In 1.9.1: 
```json
    filters={"label":["com.joyent.package=sample-2G", "foo=bar"]}
```

In 1.10.0:
```json
    filters={"label":{"com.joyent.package=sample-2G":true, "foo=bar":true}}
```

### CODE REVIEW

File: lib/backends/sdc/containers.js [hunk 1]
```diff
@@ -2309,14 +2309,26 @@ function getContainers(opts, callback) {
...
     var vmapi = opts.app.vmapi;
 
+    if (opts.filters) {
+        filters = JSON.parse(opts.filters);
+        filters = utils.getNormalizedFilters({ log: log }, filters);
+        if (filters instanceof Error) {
+            callback(new errors.DockerError('invalid filters: ' + filters));
+            return;
+        }
+        log.debug('filters: ', filters);
```
Comment (trentm) [inline lib/backends/sdc/containers.js]: Still want this? If so, please this:

`log.debug({filters: filters}, 'getContainers: filters');`

File: lib/backends/sdc/utils.js [hunk 2]
```diff
@@ -110,45 +110,46 @@ function getPublishingRules(opts, fwrules) {
...
+        } else if (typeof (val) === 'object') {
+            // Convert to an array of keys.
+            newFilters[name] = Object.keys(val);
+        } else {
+            log.warn({
+                filters: filters,
+                type: typeof (val),
+                val: JSON.stringify(val)
+            }, 'WARNING: filter %s is not an array or object', name);
+            return new Error('invalid filters');
```
Comment (trentm) [inline lib/backends/sdc/utils.js]: I'd drop the `log.warn` (I'm assuming we are going to raise the error for the response and then the audit log will show it).

Please have the Error message include details:  `return new Error(format('invalid filters: filter is not an array or object: %j', val));`

File: lib/backends/sdc/utils.js [hunk 3]
```diff
@@ -110,45 +110,46 @@ function getPublishingRules(opts, fwrules) {
...
  * In 1.10.0:
  *
- *     filters={"label":{"com.joyent.package=sample-2G":true}}
- *
- * This function takes the filters[label] value (as the filters parameter here)
- * and returns a normalized array of the packages we want to filter on.
+ *     filters={"label":{"com.joyent.package=sample-2G":true, "foo=bar":true}}
  *
+ * This function takes the filters object and returns a new normalized object
+ * with each object value being an array instance.
```
Comment (trentm) [inline lib/backends/sdc/utils.js]: I.e. the normalized form is  what it looks like in 1.9.1, right?  Might be clearer to state that or just say "the former".

File: test/lib/common.js [hunk 4]
```diff
@@ -251,6 +251,59 @@ function makeContainerName(prefix) {
 }
 
 
+/*
+ * Parse docker columnar output by using the width's in the first header row.
```
Comment (trentm) [inline test/lib/common.js]: "widths"

**Comment by trentm** [comment]
Were there tests already for filtering on *container* labels? I see one test for filtering on image labels.

**Comment by trentm** [comment]
I assume this passes the suite with each of the CLI versions (including 1.11).

LGTM.

File: lib/backends/sdc/containers.js (hunk 1 above)
Comment (twhiteman) [inline lib/backends/sdc/containers.js]: Done

File: lib/backends/sdc/utils.js (hunk 2 above)
Comment (twhiteman) [inline lib/backends/sdc/utils.js]: Done

File: lib/backends/sdc/utils.js (hunk 3 above)
Comment (twhiteman) [inline lib/backends/sdc/utils.js]: Done

File: test/lib/common.js (hunk 4 above)
Comment (twhiteman) [inline test/lib/common.js]: Done

**Comment by twhiteman** [comment]
There were no tests in sdc-docker covering container filtering - but there is good coverage in the docker/docker nightly tests.

Since the docker/docker integration tests are currently strapped to one version of docker (1.9.2 at time of writing), this is why we didn't notice this breakage.

I'll add a few cli container filter tests to sdc-docker, so that we can have *some* coverage of this across multiple docker versions.

**Comment by twhiteman** [comment]
Added container filter tests in new 'cli-filters.test.js' file.

**Pull request closed without merging**